shared/fs/row/popup-container.js

```diff
@@ -0,0 +1,89 @@
+// @flow
+import * as Types from '../../constants/types/fs'
+import * as Constants from '../../constants/fs'
+import * as FSGen from '../../actions/fs-gen'
+import {compose, connect, setDisplayName, type TypedState, type Dispatch} from '../../util/container'
+import {navigateUp} from '../../actions/route-tree'
+import Popup from './popup'
+import {fileUIName} from '../../constants/platform'
+import {isMobile} from '../../styles'
+
+const mapStateToProps = (state: TypedState, {routeProps}) => {
+  const path = routeProps.get('path')
+  const pathItem = state.fs.pathItems.get(path) || Constants.makeUnknownPathItem()
+  const _username = state.config.username || undefined
+  // We need to do this counting here since it's possible between this
```

Comment:
i don't understand this comment. this shouldn't be true

Reply:
`pathItem.children` only has names, but not their type. The race that this is addressing is like this:

1. `mapStateToProps` is called, and we have directory:

```
a dir
b dir
c file
```

2. The directory content is updated to:

```
a dir
b file
c dir
```

3. Something else triggered `folderLoad`, which caused state change.

4. `mapStateToProps` gets called again, and `pathItem.children` remains the same (`["a", "b", "c"]`) and redux skips `mergeProps`.

So if we do this counting in `mergeProps` rather than `mapStateToProps`, we wouldn't be able to capture the change that happened in step 2 because `pathItem` remained same in step 4.

Reply:
Let's talk about this over video. If the children types aren't captured that's fine but the sub connected components should then change

Reply:
Oh hmm you mean we can make "X Folders, Y Files" into its own connected component? That seems to be a great idea! Let me know if I misunderstood though.

Reply:
Oh wait I think we'd still have the same problem it's in its own connected component. We could construct an array of `PathItem`s in `mapStateToProps`, and do the actual filtering in `mergeProps`, but not sure if it's better.

So video sounds good. Thanks!